src/wallet/asyncrpcoperation_shieldcoinbase.cpp

```diff
@@ -157,303 +117,141 @@ void AsyncRPCOperation_shieldcoinbase::main() {
-        uint256 txidhash = tx_.GetHash();
-        std::shared_ptr<PaymentDisclosureDB> db = PaymentDisclosureDB::sharedInstance();
-        for (PaymentDisclosureKeyInfo p : paymentDisclosureData_) {
-            p.first.hash = txidhash;
-            if (!db->Put(p.first, p.second)) {
-                LogPrint("paymentdisclosure", "%s: Payment Disclosure: Error writing entry to database for key %s\n", getId(), p.first.ToString());
-            } else {
-                LogPrint("paymentdisclosure", "%s: Payment Disclosure: Successfully added entry to database for key %s\n", getId(), p.first.ToString());
-            }
-        }
-    }
-    // !!! Payment disclosure END
-}
-
-bool AsyncRPCOperation_shieldcoinbase::main_impl() {
-
-    CAmount minersFee = fee_;
-
-    size_t numInputs = inputs_.size();
-
-    CAmount targetAmount = 0;
-    for (ShieldCoinbaseUTXO & utxo : inputs_) {
-        targetAmount += utxo.amount;
-    }
-
-    if (targetAmount <= minersFee) {
-        throw JSONRPCError(RPC_WALLET_INSUFFICIENT_FUNDS,
-            strprintf("Insufficient coinbase funds, have %s and miners fee is %s",
-            FormatMoney(targetAmount), FormatMoney(minersFee)));
-    }
-
-    CAmount sendAmount = targetAmount - minersFee;
-    LogPrint("zrpc", "%s: spending %s to shield %s with fee %s\n",
-            getId(), FormatMoney(targetAmount), FormatMoney(sendAmount), FormatMoney(minersFee));
-
-    return std::visit(ShieldToAddress(this, sendAmount), tozaddr_);
 }
 
-void ShieldToAddress::shieldToAddress(const libzcash::RecipientAddress& recipient, AsyncRPCOperation_shieldcoinbase *m_op) {
-    m_op->builder_.SetFee(m_op->fee_);
-
-    // Sending from a t-address, which we don't have an ovk for. Instead,
-    // generate a common one from the HD seed. This ensures the data is
-    // recoverable, while keeping it logically separate from the ZIP 32
-    // Sapling key hierarchy, which the user might not be using.
-    // FIXME: update to use the ZIP-316 OVK (#5511)
-    HDSeed seed = pwalletMain->GetHDSeedForRPC();
-    uint256 ovk = ovkForShieldingFromTaddr(seed);
-
-    // Add transparent inputs
-    for (auto t : m_op->inputs_) {
-        m_op->builder_.AddTransparentInput(COutPoint(t.txid, t.vout), t.scriptPubKey, t.amount);
-    }
 
-    // Send all value to the target recipient
-    m_op->builder_.SendChangeTo(recipient, ovk);
+Remaining AsyncRPCOperation_shieldcoinbase::prepare(CWallet& wallet) {
+    auto spendable = builder_.FindAllSpendableInputs(wallet, ztxoSelector_, COINBASE_MATURITY);
 
-    // Build the transaction
-    m_op->tx_ = m_op->builder_.Build().GetTxOrThrow();
-}
+    // Find unspent coinbase utxos and update estimated size
+    unsigned int max_tx_size = MAX_TX_SIZE_AFTER_SAPLING;
+    CAmount shieldingValue = 0;
+    CAmount remainingValue = 0;
+    size_t estimatedTxSize = 10000; // per ZIP 401 (https://zips.z.cash/zip-0401#specification)
```

Comment:
```suggestion
    size_t estimatedTxSize = MIN_TX_COST; // per ZIP 401 (https://zips.z.cash/zip-0401#specification)
```
(defined in `mempool_limit.h`). But ZIP 401 doesn't specify that we use this for tx construction, so it's a bit confusing to be referencing that.

If I understand correctly, this is an initial estimate that is going to be refined in subsequent iterations, so it doesn't really matter much what we start with? In which case using MIN_TX_COST is fine, but please say something like that in the comment.

Reply:
It doesn’t really get “refined” so much as increased as new inputs are added. So we’re starting with the base being the most expensive recipient, then bump the size as we add inputs.

I’ll fix the comment (and use the constant).